### ISSUE TITLE
deep learning applications readme patch breaks tensorflow>=1.12

### ISSUE THREAD

**Issue by mbauw** [geomstats/examples#23]
Hi,

I'm trying to run geomstats deep learning applications on Ubuntu 18.04 LTS using my RTX 2080 GPU. I systematically encounter an error that suggests the patch on `variables.py` described in the repo's [readme](https://github.com/geomstats/applications/tree/master/deep_learning) breaks my tensorflow. Hardware and drivers description:

- Ubuntu LTS 18.04
- NVIDIA driver version 418.56 (proprietary)
- CUDA 10.0
- cuDNN 7.5.0 (installed from .deb files)
- RTX 2080 GPU
- Ryzen 2700x CPU

**_I have one possible explanation_**, although I find it absurd, but in case it saves us a lot of time: do the deep learning applications only work with tensorflow version 1.8 ? Because the `requirements.txt` of geomstats says `tensorflow>=1.12`, whereas the applications repo [readme](https://github.com/geomstats/applications/tree/master/deep_learning) only mentions 1.8.

I tried the following installation steps, with:

- tensorflow-gpu 1.13.1
- tf-nightly-gpu 1.14.0.dev2019032201
- tensorflow 1.13.1 (basic CPU-only version, just in case it was GPU-related in some way)

1. create a Python 3 (3.6.7) virtual env
2. install one of the three previously mentioned tensorflow and geomstats (1.15), with `pip`
3. run tensorflow [toy code](https://www.tensorflow.org/tutorials) to check if unpatched virtual env tensorflow works (it does this far in the steps, for all three versions of tf considered)
4. patch virtual env tensorflow according to the applications repo's [instructions](https://github.com/geomstats/applications/tree/master/deep_learning)
5. install patched keras thanks to the previous step instructions, doing so with a modification of the original instructions since the target is the virtual env and not the system, hence use `python setup.py install` and not the readme sudo version. The shell return allows to check which Keras was installed and updated, and more specifically the exact version and whether this was indeed done in the virtual env
```
Using /home/mbauw/Documents/virtual_env/geomstats_env_third/lib/python3.6/site-packages
Finished processing dependencies for Keras==2.1.6
```
6. try `./run.sh` or `python <MNIST or CIFAR example>.py` and encounter same error (see at the end of this issue for complete error message)
7. try tensorflow [toy code](https://www.tensorflow.org/tutorials) **_that worked in the virtual env before patching_**, and encounter same error
8. conclude that the patch broke tensorflow according to [related tf repo issue
](https://github.com/tensorflow/tensorflow/issues/20499) (they advise re-installing tensorflow to solve this error, but this basically means cancelling the patch here...)

One should note that the error specifically mentions the unique file that was modified by the patch `virtual_env/geomstats_env_third/lib/python3.6/site-packages/tensorflow/python/ops/variables.py`. 

In addition, I tried both installing Keras with pip right after tensorflow when launching the venv, and only installing Keras with the provided `setup.py` script. All ended with the same error.

## Complete error message:

```
geomstats_examples/applications/deep_learning$ python cifar_10_resnet_hypersphere.py 
Using TensorFlow backend.
Traceback (most recent call last):
  File "cifar_10_resnet_hypersphere.py", line 13, in <module>
    import keras
  File "/home/mbauw/Documents/virtual_env/geomstats_env_third/lib/python3.6/site-packages/Keras-2.1.6-py3.6.egg/keras/__init__.py", line 3, in <module>
  File "/home/mbauw/Documents/virtual_env/geomstats_env_third/lib/python3.6/site-packages/Keras-2.1.6-py3.6.egg/keras/utils/__init__.py", line 6, in <module>
  File "/home/mbauw/Documents/virtual_env/geomstats_env_third/lib/python3.6/site-packages/Keras-2.1.6-py3.6.egg/keras/utils/conv_utils.py", line 9, in <module>
  File "/home/mbauw/Documents/virtual_env/geomstats_env_third/lib/python3.6/site-packages/Keras-2.1.6-py3.6.egg/keras/backend/__init__.py", line 84, in <module>
  File "/home/mbauw/Documents/virtual_env/geomstats_env_third/lib/python3.6/site-packages/Keras-2.1.6-py3.6.egg/keras/backend/tensorflow_backend.py", line 5, in <module>
  File "/home/mbauw/Documents/virtual_env/geomstats_env_third/lib/python3.6/site-packages/tensorflow/__init__.py", line 28, in <module>
    from tensorflow.python import pywrap_tensorflow  # pylint: disable=unused-import
  File "/home/mbauw/Documents/virtual_env/geomstats_env_third/lib/python3.6/site-packages/tensorflow/python/__init__.py", line 63, in <module>
    from tensorflow.python.framework.framework_lib import *  # pylint: disable=redefined-builtin
  File "/home/mbauw/Documents/virtual_env/geomstats_env_third/lib/python3.6/site-packages/tensorflow/python/framework/framework_lib.py", line 52, in <module>
    from tensorflow.python.framework.importer import import_graph_def
  File "/home/mbauw/Documents/virtual_env/geomstats_env_third/lib/python3.6/site-packages/tensorflow/python/framework/importer.py", line 28, in <module>
    from tensorflow.python.framework import function
  File "/home/mbauw/Documents/virtual_env/geomstats_env_third/lib/python3.6/site-packages/tensorflow/python/framework/function.py", line 37, in <module>
    from tensorflow.python.ops import resource_variable_ops
  File "/home/mbauw/Documents/virtual_env/geomstats_env_third/lib/python3.6/site-packages/tensorflow/python/ops/resource_variable_ops.py", line 41, in <module>
    from tensorflow.python.ops import variables
  File "/home/mbauw/Documents/virtual_env/geomstats_env_third/lib/python3.6/site-packages/tensorflow/python/ops/variables.py", line 32, in <module>
    from tensorflow.python.training import checkpointable
  ImportError: cannot import name 'checkpointable'
```

**Comment by johmathe** [comment]
Hi, and sorry for the delay in the response,

That's right, you will need a specific version of tensorflow and a manual patch. We know this is very suboptimal and are working on a patch right now. 

Can I ask what precisely you're trying to check and which usecase? Are you trying to do optimization on a specific manifold?

Best,

Johan

**Comment by mbauw** [comment]
> Can I ask what precisely you're trying to check and which usecase? Are you trying to do optimization on a specific manifold?

If those questions are not directly related to the specific issue of this post as I understand them: I plan on using SPD matrices to do classification on signal and image processing to begin with. I'm not sure at the moment whether I will, in addition to transforming my data into SPD matrices, constrain parameters to specific manifolds, but if I manage to get geomstats to do that for me I'm interested in trying it and playing with the possibilities to see how it impacts the networks performances. Since I'm discovering Geometric DL, I'm also interested at looking at how you actually implemented the metrics etc.

To conclude this specific issue, unless I'm mistaken, your answer is simply that at the moment, constraining the parameters of a keras neural network to a chosen manifold is a geomstats feature only available using `tensorflow 1.8` then ? If yes, please excuse the issue opening, since it's kinda explained in the application readme.

In any case thanks for the lib' & for your kind answer :+1:

**Comment by johmathe** [comment]
I think that in your case you don't need to use this example as a base - you should probably be looking at the SPD application example.

My understanding is that your manifold is on the activations, not the weights of the network.

You are right that this example is currently only for tensorflow 1.8. Indeed if we wanted to have this in all the tensorflow we would need to send the patch upstream or actively maintain it which would probably be a full time job. We are open to contributions or ideas if you think that there are better ways to keep this working in all versions.

Thanks for the detailed bug report and the questions - we appreciate it :)
